File: src/main/java/me/testaccount666/serversystem/clickablesigns/executables/give/ConfiguratorGiveSign.java
package me.testaccount666.serversystem.clickablesigns.executables.give;

import me.testaccount666.serversystem.ServerSystem;
import me.testaccount666.serversystem.clickablesigns.AbstractSignConfigurator;
import me.testaccount666.serversystem.clickablesigns.SignManager;
import me.testaccount666.serversystem.clickablesigns.SignType;
import me.testaccount666.serversystem.clickablesigns.cost.CostType;
import me.testaccount666.serversystem.clickablesigns.util.SignUtils;
import me.testaccount666.serversystem.managers.PermissionManager;
import me.testaccount666.serversystem.userdata.User;
import me.testaccount666.serversystem.utils.BiDirectionalHashMap;
import me.testaccount666.serversystem.utils.ComponentColor;
import org.bukkit.block.Sign;
import org.bukkit.block.sign.Side;
import org.bukkit.configuration.file.FileConfiguration;
import org.bukkit.configuration.file.YamlConfiguration;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.block.Action;
import org.bukkit.event.player.PlayerInteractEvent;

import java.io.IOException;
import java.util.logging.Level;

import static me.testaccount666.serversystem.utils.MessageBuilder.general;
import static me.testaccount666.serversystem.utils.MessageBuilder.sign;

public class ConfiguratorGiveSign extends AbstractSignConfigurator implements Listener {
    private static final BiDirectionalHashMap<User, Sign> _CONFIGURATORS = new BiDirectionalHashMap<>();

    @Override
    protected String getCreatePermissionNode() {
        return "ClickableSigns.Give.Create";
    }

    @Override
    protected SignType getSignType() {
        return SignType.GIVE;
    }

    @Override
    protected boolean validateConfiguration(User user, Sign sign, YamlConfiguration config) {
        return true;
    }

    @Override
    protected void addSignSpecificConfiguration(User user, Sign sign, FileConfiguration config) {
    }

    @Override
    protected String getSuccessMessageKey() {
        return "Give.Created";
    }

    @Override
    public void execute(User user, Sign sign) {
        if (!validatePermission(user)) return;

        _CONFIGURATORS.put(user, sign);
        sign("Give.Configuring", user).build();
    }

    @EventHandler
    public void onSignRightClick(PlayerInteractEvent event) {
        if (event.getAction() != Action.RIGHT_CLICK_BLOCK) return;
        var clickedBlock = event.getClickedBlock();
        if (clickedBlock == null) return;
        if (!(clickedBlock.getState() instanceof Sign sign)) return;
        var userOptional = _CONFIGURATORS.getKey(sign);
        if (userOptional.isEmpty()) return;
        var user = userOptional.get();
        if (!user.getUuid().equals(event.getPlayer().getUniqueId())) return;
        event.setCancelled(true);

        var itemToGive = event.getPlayer().getInventory().getItemInMainHand();
        if (itemToGive.isEmpty()) {
            sign("Give.NoItem", user).build();
            return;
        }

        var signFile = SignUtils.getSignFile(sign.getLocation());
        var config = YamlConfiguration.loadConfiguration(signFile);

        config.set("Key", getSignType().name());

        config.set("Cost.Type", CostType.NONE.name());
        config.set("Cost.Amount", 0);

        config.set("Item", itemToGive);

        try {
            config.save(signFile);
        } catch (IOException exception) {
            user.sendMessage(exception.getMessage());
            ServerSystem.getLog().log(Level.SEVERE, "Failed to save sign configuration ${signFile.getAbsolutePath()}", exception);
            return;
        }

        ServerSystem.getInstance().getRegistry().getService(SignManager.class).addSignType(sign.getLocation(), getSignType());
        _CONFIGURATORS.removeByValue(sign);

        var front = sign.getSide(Side.FRONT);
        front.line(0, ComponentColor.translateToComponent(SignType.GIVE.signName()));
        front.line(1, ComponentColor.translateToComponent("&2${itemToGive.getType().name()}"));
        var back = sign.getSide(Side.BACK);
        for (var index = 0; index < 4; index++) back.line(index, front.line(index));
        sign.update();
    }

    /**
     * Validates that the user has permission to create this sign.
     *
     * @param user The user to check
     * @return true if the user has permission, false otherwise
     */
    private boolean validatePermission(User user) {
        if (!PermissionManager.hasPermission(user, getCreatePermissionNode(), false)) {
            general("NoPermission", user)
                    .postModifier(message -> message.replace("<PERMISSION>",
                            PermissionManager.getPermission(getCreatePermissionNode()))).build();
            return false;
        }
        return true;
    }
}

File: src/main/kotlin/me/testaccount666/serversystem/clickablesigns/executables/give/ConfiguratorGiveSign.kt
package me.testaccount666.serversystem.clickablesigns.executables.give

import me.testaccount666.paperktx.extensions.ComponentExtensions.asComponent
import me.testaccount666.paperktx.extensions.isAir
import me.testaccount666.serversystem.ServerSystem.Companion.log
import me.testaccount666.serversystem.clickablesigns.*
import me.testaccount666.serversystem.clickablesigns.cost.CostType
import me.testaccount666.serversystem.clickablesigns.util.SignUtils
import me.testaccount666.serversystem.extensions.*
import me.testaccount666.serversystem.managers.PermissionManager.getPermission
import me.testaccount666.serversystem.managers.PermissionManager.hasPermission
import me.testaccount666.serversystem.userdata.User
import me.testaccount666.serversystem.utils.BiDirectionalHashMap
import org.bukkit.block.Sign
import org.bukkit.block.sign.Side
import org.bukkit.configuration.file.FileConfiguration
import org.bukkit.configuration.file.YamlConfiguration
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.block.Action
import org.bukkit.event.player.PlayerInteractEvent
import java.io.IOException
import java.util.logging.Level

class ConfiguratorGiveSign : AbstractSignConfigurator(), Listener {
    override val createPermissionNode = "ClickableSigns.Give.Create"
    override val successMessageKey = "Give.Created"
    override val signType = SignType.GIVE

    override fun validateConfiguration(user: User, sign: Sign, config: YamlConfiguration) = true

    override fun addSignSpecificConfiguration(user: User, sign: Sign, config: FileConfiguration) {
    }

    override fun execute(user: User, sign: Sign) {
        if (!validatePermission(user)) return

        _CONFIGURATORS.put(user, sign)
        user.signMsg("Give.Configuring")
    }

    @EventHandler
    fun onSignRightClick(event: PlayerInteractEvent) {
        if (event.action != Action.RIGHT_CLICK_BLOCK) return

        val clickedBlock = event.clickedBlock ?: return
        val sign = clickedBlock.state as? Sign ?: return

        val user = _CONFIGURATORS.getKey(sign) ?: return
        if (user.uuid != event.getPlayer().uniqueId) return
        event.isCancelled = true

        val itemToGive = event.getPlayer().inventory.itemInMainHand
        if (itemToGive.isAir()) {
            user.signMsg("Give.NoItem")
            return
        }

        val signFile = SignUtils.getSignFile(sign.location)
        val config = YamlConfiguration.loadConfiguration(signFile)

        config["Key"] = signType.name

        config["Cost.Type"] = CostType.NONE.name
        config["Cost.Amount"] = 0

        config["Item"] = itemToGive

        try {
            config.save(signFile)
        } catch (exception: IOException) {
            user.sendMessage(exception.message ?: "null")
            log.log(Level.SEVERE, "Failed to save sign configuration ${signFile.absolutePath}", exception)
            return
        }

        getService<SignManager>().addSignType(sign.location, signType)
        _CONFIGURATORS.removeByValue(sign)

        val front = sign.getSide(Side.FRONT)
        front.line(0, SignType.GIVE.signName.asComponent())
        front.line(1, "&2${itemToGive.type.name}".asComponent())
        val back = sign.getSide(Side.BACK)
        for (index in 0..3) back.line(index, front.line(index))
        sign.update()
    }

    /**
     * Validates that the user has permission to create this sign.
     *
     * @param user The user to check
     * @return true if the user has permission, false otherwise
     */
    private fun validatePermission(user: User): Boolean {
        if (!hasPermission(user, createPermissionNode, false)) {
            user.generalMsg("NoPermission") {
                postModifier {
                    it.replace("<PERMISSION>", getPermission(createPermissionNode)!!)
                }
            }
            return false
        }
        return true
    }

    companion object {
        private val _CONFIGURATORS = BiDirectionalHashMap<User, Sign>()
    }
}